docker_container - adding publish_all_ports option

Ajpantuso opened the pull request:
##### SUMMARY
Adding `publish_all_ports` option to accurately reflect Docker CLI behavior.
Fixes #8
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME
plugins/modules/docker_container.py
##### ADDITIONAL INFORMATION
The Docker API and Docker CLI combine the use of `ports` and `publish_all_ports` similarly, but the module logic was treating these as mutually exclusive values. This change simply splits the options such that the old way of publishing all ports is available until deprecation, but the new way will work as intended.

The following task:
```yaml
 - hosts: localhost
   collections:
     - community.docker
   tasks:
     - docker_container:
         name: "nginx"
         image: nginx
         pull: yes
         publish_all_ports: true
         ports:
           - 8080:8080
         state: started
       become: true
```
now results in:
```bash
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                                           NAMES
94798f1bf0bf        nginx               "/docker-entrypoint.…"   4 seconds ago       Up 3 seconds        0.0.0.0:8080->8080/tcp, 0.0.0.0:32793->80/tcp   nginx
```

felixfontein commented:
You also have to update the deprecation message: https://github.com/ansible-collections/community.docker/blob/18e6bd20d26382ddc94680aecdb3f51df36180a1/plugins/modules/docker_container.py#L1757-L1763.

Also using `all` in `published_ports` should not be allowed if `publish_all_ports` is set to `true`.

Ajpantuso commented:
> You also have to update the deprecation message:

Done.
> Also using `all` in `published_ports` should not be allowed if `publish_all_ports` is set to `true`.

Module will now fail in this scenario.

felixfontein commented:
Cool, thanks!

Can you add a test to tests/integration/targets/docker_container/tasks/tests/ports.yml which makes sure that `published_ports: all` is equivalent to `publish_all_ports=true`? (i.e. one more task in the `published_ports: all` series)

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -663,6 +663,13 @@
+      - Publish all ports to the host.
+      - Any specified port bindings from I(published_ports) will remain intact when C(true).
+    type: bool
+    default: false
```
Right now I'm not really happy with the default for `publish_all_ports`. This is only partially a problem with this PR though, since that behavior was already there before.

The problem is that right now, if the user does not specify `published_ports`, the published ports are not checked for idempotence - **but** the `publish_all_ports` flag is! (If a container has it set, and `docker_container`'s `published_ports` isn't specified, the container will be recreated since there's a config mismatch!) I would argue that the current behavior is a bug, but potentially people depend on it...

The *correct* behavior would be (before this PR) to set `publish_all_ports = None`, and only set it to `True`/`False` if `published_ports` is passed to the module. With the new option it get's a bit more complicated, but it should also not be a problem.

The main problem is what do we do about the old problem. I'm currently tending to treat it as a bug and simply fix it. This changes behavior, but since it prevents container recreation in some situations, I think it's more acceptable than it would be the other way around. What do you think?

Ajpantuso replied:
I tend to agree with you. Any predictable side-effect could be misunderstood as a feature, but the module already has a feature (`recreate`) to let users force recreation in a more obvious way if desired. In other words if they really needed the container to be recreated they should use the intended option otherwise the module is falsely reporting config changes which is a bug.

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -1424,11 +1431,14 @@ def __init__(self, client):
         if self.published_ports == 'all':
-            self.publish_all_ports = True
-            self.published_ports = None
+            if not self.publish_all_ports:
```
```suggestion
            if self.publish_all_ports is None:
```

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -1424,11 +1431,14 @@ def __init__(self, client):
+                self.publish_all_ports = True
+                self.published_ports = None
+            else:
+                self.fail('"all" is not a valid value for "published_ports" when "publish_all_ports" is "true"')
```
```suggestion
                self.fail('"all" is not a valid value for "published_ports" when "publish_all_ports" is specified')
```

felixfontein reviewed (commented):
I think we should also have a set of integration tests for idempotency (and changed status) of `publish_all_ports` combined with `published_ports`.

Ajpantuso commented:
> I think we should also have a set of integration tests for idempotency (and changed status) of `publish_all_ports` combined with `published_ports`.

Not sure if I understood fully, but I added tests that cover transitions between different combinations of these options.

felixfontein commented on plugins/modules/docker_container.py:
```diff
@@ -1750,10 +1759,9 @@ def _parse_publish_ports(self):
-                    'currently use the module will change in a breaking way',
+                    'supported by the module. The port mappings are currently ignored. Set publish_all_ports '
+                    'to "true" to randomly assign port mappings for those not specified by published_ports. '
+                    'The use of "all" in published_ports will be removed in version 2.0.0.',
```
```suggestion
                    'The use of "all" in published_ports next to other values will be removed in version 2.0.0.',
```
`all` will still be allowed in 2.0.0. What we will do in 2.0.0 is deprecate `all` (with removal in 3.0.0). But that shouldn't be mentioned here.

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/ports.yml:
```diff
@@ -284,3 +298,154 @@
+    command: '/bin/sh -c "sleep 10m"'
+    name: "{{ cname }}"
+    state: started
+    publish_all_ports: false
```
```suggestion
```

felixfontein reviewed (commented):
Right now, you're only testing `publish_all_ports` not specified after `publish_all_ports: false`. You must also test `publish_all_ports` not specified after `publish_all_ports: true`, since that's the actually interesting and important case.

Ajpantuso commented:
> Right now, you're only testing `publish_all_ports` not specified after `publish_all_ports: false`. You must also test `publish_all_ports` not specified after `publish_all_ports: true`, since that's the actually interesting and important case.

Ok, made the tests easier to read this time and 'true to unspecified' is now covered.

felixfontein commented on tests/integration/targets/docker_container/tasks/tests/ports.yml:
```diff
@@ -303,149 +303,78 @@
-      - publish_all_ports_10 is not changed
-      - publish_all_ports_11 is not changed
+      - publish_all_ports.results[index].changed == publish_all_ports_test_cases[index].changed
+  loop: "{{ range(0, publish_all_ports_test_cases | length) | list }}"
```
For things like this, the `indexed_items` lookup is also pretty useful (https://docs.ansible.com/ansible/latest/collections/ansible/builtin/indexed_items_lookup.html#ansible-collections-ansible-builtin-indexed-items-lookup).

Ajpantuso replied:
Thanks, I updated it to the `loop` analouge.

felixfontein commented:
@Ajpantuso thanks a lot for implementing this!
